src/sources/kafka.rs

```diff
@@ -460,10 +490,48 @@ struct KafkaPartitionState {
     /// can complete before we get a rebalance callback, so "observed complete" (based on seeing the end of the stream)
     /// and "expect to complete" (based on seeing a rebalance callback with revoked partition info) are tracked separately.
     observed_drain: HashSet<TopicPartition>,
```

Comment:
I've been going back and forth on this question tbh. It's a case that can only happen in that "reached end of partition" mode, because that is the only condition where the consumer tasks will exit without also being added to the "expected drain" set. The comment on this field is still technically correct, but really reflects an earlier iteration where the callback handler closed the finalizer stream _before_ sending the rebalance callback, so observing a drained partition prior to populating it in the "expected to drain" set was a regular occurrence.

Since this EOF mode is only used in the tests, I'm a little hesitant to add code for it that would always be enabled. However, the "worst case" scenario is something like:

1. the consumer is in "exit at EOF" mode
2. and one or more partitions finish, while others are still being consumed
3. a rebalance starts and some of the already-done partitions are being revoked

The drain deadline would take effect in this case, since the already-finished tasks can no longer be observed. So, it won't deadlock, but the rebalance will likely take longer than it otherwise would.

There are a couple ways to address this: 

- explicitly track a set of partitions that end during the consuming phase on the `Consuming` struct, and pass that along to the `Draining` struct
- Or, when populating the set of partitions expected to drain, only add the partitions that had an entry in `end_signals`

I'm inclined to take latter approach, since it should solve this case without adding any overhead (although, it shouldn't be too expensive to explicitly tracking these, either -- mostly it would mean allocating another `Set` that is rarely, if ever, used).

Reply:
OK. there's a case to be made here that `observed_drain` is not needed at all anymore: removing entries from `expect_drain` as tasks complete, and checking if that set is empty, is enough to detect the end of the draining phase now. The guarantees that come with 1) handling callbacks and the partition task completions on the same task, and 2) keeping the ack stream and message reading for a given partition together in the same task, and only exiting after all acks are drained, mean that this should be safe.

In fact...thinking through the case where the deadline kicks in and some tasks get forcibly ended, that is a scenario that _could_ end up adding partitions to `observed_drain` during the consuming phase, and keeping that state could be a problem. An [aborted task _may_ still return `Ok`](https://docs.rs/tokio/latest/tokio/task/struct.AbortHandle.html#method.abort) after being aborted, and if that happens, the following could play out:

1. The state machine is in the consuming phase after aborting the previous drain due to timeout
2. One of the partition consumer tasks is seen to complete, and added to the `observed_drain` set
4. Partitions are assigned, including the one we just observed complete (this may happen before or after 2, so clearing `observed_drain` during assignment doesn't help)
5. Later, another rebalance is initiated and this partition is revoked again. Now the draining phase is at risk of completing before this partition is actually finished the second time

To correctly track "observed drain" through the consuming phase, we would need to introduce something like a generation counter to prevent this being a problem. However, for now I think we can actually remove `observed_drain` safely, provided `expect_drain` is only populated with partitions that are known to still have a running (or complete but not yet observed via `join_next`) task.

Reply:
Very true, I agree with your assessment. Thanks for thinking through that and the detailed explanation!